### DISCUSSION TITLE
Performance data on thousands of EC2s in Reference Architecture

### DISCUSSION THREAD

**Discussion by iantybo** [gruntwork-io/knowledge-base#480] [Help] [answered]
Since the Reference Architecture is based on Terragrunt, does this buy us any performance gains as far as `plan`s and `apply`s go for workloads that consist of thousands of EC2s? What are some considerations I need to make while I design my modules?

---

<ins datetime="2022-06-22T14:01:25Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108812">Tracked in ticket #108812</a></p>
</ins>

**Comment by yorinasub17** [comment, marked as answer]
This depends on how the deployment is structured.

If your EC2 instances are grouped across multiple apps that each have different deployment schedules, then Terragrunt can certainly help by allowing you to better break down your deployments so that you are only deploying to a subset of the resources at a time.

If instead all those EC2 instances are for a single app/cluster such that you need to roll out to all of them at once each time, then Terragrunt doesn't really help because those will be managed by a single ASG/terraform module and thus the features of Terragrunt don't apply.

In general, Terragrunt is a tool to make it much easier to manage multi-state file Terraform projects, so it will only help if you can break down your infrastructure into multiple components that have different deployment schedules.

Refer to https://github.com/gruntwork-io/knowledge-base/discussions/402 for an extended commentary on this topic.
